### PR TITLE
refactor(core): avoid creating a DOM node before creating a TNode

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit updates the `elementStart` instruction to avoid creating DOM nodes before creating a corresponding TNode. This refactoring is needed to make sure this internal logic is consistent across all instructions.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/instructions/shared.ts [hunk 1]
```diff
@@ -986,12 +986,11 @@ export function setNgReflectProperties(
  */
 export function resolveDirectives(
     tView: TView, lView: LView, tNode: TElementNode|TContainerNode|TElementContainerNode,
-    localRefs: string[]|null): boolean {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/shared.ts:989]: Question: is this change needed as part of the refactoring? 
I don't mind this change, just curious about the motivation.

**Review by pkozlowski-opensource — COMMENTED** [review]
@AndrewKushnir LGTM, just wanted to check with you regarding the `resolveDirectives` signature

File: packages/core/src/render3/instructions/shared.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/shared.ts:989]: The `resolveDirectives` call sites do not use its return value. There is also a special flag on the TNode (the `TNodeFlags.isDirectiveHost` one), so returning the "has directives" value is not strictly necessary.

Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/shared.ts:989]: Hm, the `elementStartFirstCreatePass` seem to use this return value. But I agree with you that it could be replaced with `TNodeFlags.isDirectiveHost` check.

Thnx for clarifying.

**Review by pkozlowski-opensource — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
Caretaker note: the presubmit is all "green" except 1 flaky test.

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 02f4599669a22d6af5a3e1a6fb1859b751a77b1b.
